src/cephadm/cephadm

```diff
@@ -1154,6 +1155,28 @@ def default_image(func):
 
     return _default_image
 
+class container_registry_credentials(ContextDecorator):
+    def __init__(self):
+        if not args.registry_username:
+            self.do_login = False
+        elif not args.registry_password:
+            self.do_login = False
+        else:
+            self.do_login = True
+
+    def __enter__(self):
+        if self.do_login:
+            out, err, code = call([container_path, 'login',
+                                   '-u', args.registry_username,
+                                   '-p', args.registry_password])
+            if code != 0:
+                logger.warning('unable to login: %s', err)
+                self.do_login = False
+
+    def __exit__(self):
+        if self.do_login:
+            call([container_path, 'logout'])
```

Comment:
just wondering. would something like this also work?

```python
def container_registry_credentials(func):
    do_login = ...
    def wrapper(*args, *+kwargs):
        try:
            ... = call([container_path, 'login', ...
            func(*args, **kwargs)
        finally:
            call([container_path, 'logout'])
    return wrapper
```
?

Reply:
This is actually my first version. Was using context manager. I think it’d be nice to use it as a decorator. 

If you think it’s an overdesign. I will use a this approach.